### PR TITLE
util: Add `rng` utilities

### PR DESCRIPTION
**Pull request by LucioFranco**
This adds new PRNG utilities that only use libstd and not the external
`rand` crate. This change's motivation are that in tower middleware that
need PRNG don't need the complexity and vast utilities of the `rand`
crate.

This adds a `Rng` trait which abstracts the simple PRNG features tower
needs. This also provides a `HasherRng` which uses the `RandomState`
type from libstd to generate random `u64` values. In addition, there is
an internal only `sample_inplace` which is used within the balance p2c
middleware to randomly pick a ready service. This implementation is
crate private since its quite specific to the balance implementation.

The goal of this in addition to the balance middlware getting `rand`
removed is for the upcoming `Retry` changes. The `next_f64` will be used
in the jitter portion of the backoff utilities in https://github.com/tower-rs/tower/pull/685.

### CODE REVIEW

File: tower/src/util/rng.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        HasherRng { hasher, counter: 0 }
+    }
+}
+
+impl<H> Rng for HasherRng<H>
+where
+    H: BuildHasher,
+{
+    fn next_u64(&mut self) -> u64 {
+        let mut hasher = self.hasher.build_hasher();
```
Comment (seanmonstar) [inline tower/src/util/rng.rs:105]: I don't think you want to `build_hasher` for each number. That will restart the hasher back to the beginning. I'd expect you'd build a hasher once, and then feed the counter into it over and over. Or am I wrong?

File: tower/src/util/rng.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
+//! [PRNG] utilities for tower middleware.
+//!
+//! This module provides a generic [`Rng`] trait and a [`HasherRng`] that
+//! implements the trait based on [`RandomState`] or any other [`Hasher`].
+//!
+//! These utlities replace tower's internal usage of `rand` with these smaller
+//! more light weight methods. Most of the implemenations are extracted from
```
Comment (hawkw) [inline tower/src/util/rng.rs:7]:
```suggestion
//! more lightweight methods. Most of the implementations are extracted from
```

File: tower/src/util/rng.rs [hunk 3]
```diff
@@ -0,0 +1,198 @@
+//! [PRNG] utilities for tower middleware.
+//!
+//! This module provides a generic [`Rng`] trait and a [`HasherRng`] that
+//! implements the trait based on [`RandomState`] or any other [`Hasher`].
+//!
+//! These utlities replace tower's internal usage of `rand` with these smaller
```
Comment (hawkw) [inline tower/src/util/rng.rs:6]:
```suggestion
//! These utlities replace tower's internal usage of `rand` with these smaller,
```

File: tower/src/util/rng.rs (hunk 1 above)
Comment (LucioFranco) [reply tower/src/util/rng.rs:105]: I thought finished consumed it, but I guess finish just takes a &self. Let me see if that makes any meaningful difference.

Comment (seanmonstar) [reply tower/src/util/rng.rs:105]: Yea, thinking more, I could be much more wrong than right.

**Review by seanmonstar — APPROVED** [review]

File: tower/src/util/rng.rs (hunk 1 above)
Comment (LucioFranco) [reply tower/src/util/rng.rs:105]: Follow up here HashMap creates a hasher for each hash so we are reverting back to that setup.

**Pull request merged**